pkg/tfbridge/detailed_diff.go

```diff
@@ -0,0 +1,406 @@
+	}
+	slices.Sort(keysSlice)
+	return keysSlice
+}
+
+func promoteToReplace(diff *pulumirpc.PropertyDiff) *pulumirpc.PropertyDiff {
+	if diff == nil {
+		return nil
+	}
+
+	kind := diff.GetKind()
+	switch kind {
+	case pulumirpc.PropertyDiff_ADD:
+		return &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_ADD_REPLACE}
+	case pulumirpc.PropertyDiff_DELETE:
+		return &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_DELETE_REPLACE}
+	case pulumirpc.PropertyDiff_UPDATE:
+		return &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_UPDATE_REPLACE}
+	default:
+		return diff
+	}
+}
+
+type baseDiff string
+
+const (
+	NoDiff    baseDiff = "NoDiff"
+	Add       baseDiff = "Add"
+	Delete    baseDiff = "Delete"
+	Update    baseDiff = "Update"
+	Undecided baseDiff = "Undecided"
+)
+
+func (b baseDiff) ToPropertyDiff() *pulumirpc.PropertyDiff {
+	contract.Assertf(b != Undecided, "diff should not be undecided")
+	switch b {
+	case Add:
+		return &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_ADD}
+	case Delete:
+		return &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_DELETE}
+	case Update:
+		return &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_UPDATE}
+	default:
+		return nil
+	}
+}
+
+func makeBaseDiff(old, new resource.PropertyValue) baseDiff {
+	oldPresent := isPresent(old)
+	newPresent := isPresent(new)
+	if !oldPresent {
+		if !newPresent {
+			return NoDiff
+		}
+
+		return Add
+	}
+	if !newPresent {
+		return Delete
+	}
+
+	if new.IsComputed() {
+		return Update
+	}
+
+	return Undecided
+}
+
+type (
+	detailedDiffKey string
+	propertyPath    resource.PropertyPath
+)
+
+func newPropertyPath(root resource.PropertyKey) propertyPath {
+	return propertyPath{string(root)}
+}
+
+func (k propertyPath) String() string {
+	return resource.PropertyPath(k).String()
+}
+
+func (k propertyPath) Key() detailedDiffKey {
+	return detailedDiffKey(k.String())
+}
+
+func (k propertyPath) append(subkey interface{}) propertyPath {
+	return append(k, subkey)
+}
+
+func (k propertyPath) Subpath(subkey string) propertyPath {
+	return k.append(subkey)
+}
+
+func (k propertyPath) Index(i int) propertyPath {
+	return k.append(i)
+}
+
+func (k propertyPath) IsReservedKey() bool {
+	leaf := k[len(k)-1]
+	return leaf == "__meta" || leaf == "__defaults"
+}
+
+func mapHasReplacements(m map[detailedDiffKey]*pulumirpc.PropertyDiff) bool {
+	for _, diff := range m {
+		if diff.GetKind() == pulumirpc.PropertyDiff_ADD_REPLACE ||
+			diff.GetKind() == pulumirpc.PropertyDiff_DELETE_REPLACE ||
+			diff.GetKind() == pulumirpc.PropertyDiff_UPDATE_REPLACE {
+			return true
+		}
+	}
+	return false
+}
+
+type detailedDiffer struct {
+	tfs shim.SchemaMap
+	ps  map[string]*SchemaInfo
+}
+
+func (differ detailedDiffer) propertyPathToSchemaPath(path propertyPath) walk.SchemaPath {
+	return PropertyPathToSchemaPath(resource.PropertyPath(path), differ.tfs, differ.ps)
+}
+
+func (differ detailedDiffer) getEffectiveType(path walk.SchemaPath) shim.ValueType {
```

Comment:
I've pondered moving this somewhere  out of this module and it's a bit tricky. I think it's valid though. It's trying to describe the Pulumi type that will be generated for the current schema pairs. shim.ValueType return type is misleading here as we're trying to describe the Pulumi type not the TF type, but we lack a good enumeration in the codebase. Perhaps just wrap it for commentary. Inside the function I think the impl is OK: either flattening is happening per IsMaxItemsOne, then the Pulumi type will correspond to the Elem(); or it's not happening, in which case objects can be recognized by TypeMap and Elem()=*Resource. Looks like caller does not distinguish Map and Object so it works out.

Reply:
I've added tests https://github.com/pulumi/pulumi-terraform-bridge/pull/2456 for all the cases I could think of.

Reply:
TypeMap and Elem() = *Resource is not a valid case - it does not pass the schema validation, so providers can not specify that